sensing/pointcloud_preprocessor/include/pointcloud_preprocessor/outlier_filter/ring_outlier_filter_nodelet.hpp

```diff
@@ -45,18 +54,27 @@ class RingOutlierFilterComponent : public pointcloud_preprocessor::Filter
   rcl_interfaces::msg::SetParametersResult paramCallback(const std::vector<rclcpp::Parameter> & p);
 
   bool isCluster(
-    sensor_msgs::msg::PointCloud2::SharedPtr & input_ptr,
-    const std::vector<std::size_t> & tmp_indices)
+    const PointCloud2ConstPtr & input, std::pair<int, int> data_idx_both_ends, int walk_size,
+    const TransformInfo & transform_info)
   {
-    PointXYZI * front_pt = reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_indices.front()]);
-    PointXYZI * back_pt = reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_indices.back()]);
-
-    const auto x_diff = front_pt->x - back_pt->x;
-    const auto y_diff = front_pt->y - back_pt->y;
-    const auto z_diff = front_pt->z - back_pt->z;
-    return static_cast<int>(tmp_indices.size()) > num_points_threshold_ ||
-           (x_diff * x_diff) + (y_diff * y_diff) + (z_diff * z_diff) >=
-             object_length_threshold_ * object_length_threshold_;
+    if (walk_size > num_points_threshold_) return true;
+
+    auto first_point = reinterpret_cast<const PointXYZI *>(&input->data[data_idx_both_ends.first]);
+    auto last_point = reinterpret_cast<const PointXYZI *>(&input->data[data_idx_both_ends.second]);
+
+    Eigen::Vector4f p1(first_point->x, first_point->y, first_point->z, 1);
+    Eigen::Vector4f p2(last_point->x, last_point->y, last_point->z, 1);
+
+    if (transform_info.need_transform) {
+      p1 = transform_info.eigen_transform * p1;
+      p2 = transform_info.eigen_transform * p2;
+    }
+
+    auto x = p1[0] - p2[0];
+    auto y = p1[1] - p2[1];
+    auto z = p1[2] - p2[2];
+
+    return x * x + y * y + z * z >= object_length_threshold_ * object_length_threshold_;
```

Comment:
Further optimization idea:
Another way to compute the distance between the 2 points is to use the [law of cosines](https://en.wikipedia.org/wiki/Law_of_cosines): `dist2(p1,p2) = d1^2 + d2^2 - 2*d1*d2*cos(a)`. Where `d1` and `d2` are the `distance` attribute of `p1` and `p2`, and `a` the `azimut` diff between the 2 points.

What is interesting with this approach is that `a` is always small. Indeed, since `isCluster` has early returns when walk has more than `num_points_threshold_` (3 by default), it means the current walk has 3 points or less. But to be added to the walk, the azimut diff between 2 points must be [less than 1 degree](https://github.com/autowarefoundation/autoware.universe/blob/7213591028bde203b6587b428db51b2159de01ff/sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp#L115). So the `a` value is at most 3 degrees. Thus `cos(a) ~= 1`.

The distance calculation can be simplified to: `dist2(p1,p2) = (d1 - d2)^2`. 

-> we don't need to fetch the xyz points anymore, just the 2 `distance` attributes are necessary.

Reply:
This proposal changes the logical output by approximation. It is beyond the scope of this Pull Request to speed up the processing without changing the logical output. It would be appreciated if you could start a new issue for discussion and agreement.